feat(pathlike): support pathlike throughout public interface

w-bonelli opened the pull request:
Working toward #1729, continuation of #1712, #1603, and #1583

- update code where needed to support `os.Pathlike` in user-facing API
- update docstrings and type hints
- update most tests to use `Path` instead of `str`
- add tests checking both `Path` and `str` for core classes/utilities
- add`verbose` flag to more utility methods to optionally enable/suppress output
- update some notebooks to use `pathlib`, but not all
- add `sim_path` property to `MFSimulation` as a shortcut to `simulation_data.mfpath.get_sim_path()`
  - guaranteed `Path` whether user provides workspace as `str` or `PathLike`
  - update tests and mf6 tutorial notebook to exercise/demonstrate
- preserve `str` properties where this was already the convention
  - e.g. `BaseModel.model_ws`
  - could consider refactoring to `Path` in future, but may break usages
- update `resolve_exe` function, add tests, and use it from both `BaseModel` as well as `run_model()`

mwtoews reviewed (commented):
Thanks, looks good so far, see following comments and suggestions. Are there a few representative tests that use `str` remaining?

w-bonelli commented on flopy/mbase.py:
```diff
@@ -866,20 +867,20 @@ def remove_output(self, fname=None, unit=None):
 
-    def get_output(self, fname=None, unit=None):
+    def get_output(
+        self, fname: Union[str, os.PathLike, None] = None, unit=None
```
Thanks, switched to `Optional[Union[str, os.PathLike]]` (need the inner union)

w-bonelli commented:
Added/expanded a few tests covering both `str` and `Path` for `Modflow` and `MFSimulation` read/write, gridgen, some export utilities. Planning to add a few more.

mwtoews commented on flopy/mbase.py:
```diff
@@ -830,20 +833,19 @@ def add_output(self, fname, unit, binflag=False, package=None):
+        unit : int, optional
+            Unit number of output array
         """
         if fname is not None:
```
probably add `fname = str(fname)` after here

mwtoews commented on flopy/mbase.py:
```diff
@@ -866,20 +868,20 @@ def remove_output(self, fname=None, unit=None):
+        unit : int, optional
+            Unit number of output array
         """
         if fname is not None:
```
also add `fname = str(fname)` after here

mwtoews commented on flopy/mbase.py:
```diff
@@ -894,21 +896,24 @@ def get_output(self, fname=None, unit=None):
+            Unit number of output array
         """
         idx = None
         if fname is not None:
```
also add `fname = str(fname)` after here

mwtoews commented on flopy/mbase.py:
```diff
@@ -787,9 +789,10 @@ def add_output_file(
+    ):
         """
         Assign an external array so that it will be listed as a DATA or
         DATA(BINARY) entry in the name file.  This will allow an outside
```
add `fname = str(fname)` below

mwtoews commented on flopy/mbase.py:
```diff
@@ -724,26 +723,29 @@ def _output_msg(self, i, add=True):
+            The package the output file is attached to, default is None
 
         """
         add_cbc = False
```
add `fname = str(fname)` after `if fname is None:` block

mwtoews commented on flopy/mbase.py:
```diff
@@ -1656,16 +1672,16 @@ def run_model(
-    namefile : str
-        Namefile of model to run. The namefile must be the
+    exe_name : str or PathLike
+        Executable name (or path, if necessary) to run.
```
before `exe_name.lower()` on L1722 be sure to use `exe_name = str(exe_name)`

mwtoews commented on flopy/mf6/mfbase.py:
```diff
@@ -201,7 +202,7 @@ class MFFileMgmt:
     """
 
-    def __init__(self, path, mfsim=None):
+    def __init__(self, path: Union[str, os.PathLike], mfsim=None):
```
above on L194 adjust docstring with `str or PathLike`

mwtoews reviewed (commented):
Nice stuff, getting closer, see attached suggestions/comments. I'm primarily looking at the main module, and have not seen many of the edits to notebooks or autotests.

langevin-usgs commented on flopy/export/netcdf.py:
```diff
@@ -152,6 +153,10 @@ def __init__(
         **kwargs,
     ):
+        if isinstance(output_filename, os.PathLike):
+            output_filename = str(
+                Path(output_filename).expanduser().absolute()
```
Out of curiosity, why is `str` needed here?

w-bonelli replied:
it's not strictly &mdash; previously `output_filename` was a `str`, switching to `Path` would require changes to L160 and L523 and I figured it may be safer to keep it a string in case any external usages had similar expectations

langevin-usgs reviewed (approved):
Looks great @w-bonelli.  Nice addition.  As you mentioned, a few remaining os.path.join remain in notebooks, but we can get those later.

w-bonelli commented on flopy/mbase.py:
```diff
@@ -693,7 +688,11 @@ def __getattr__(self, item):
     def get_ext_dict_attr(
-        self, ext_unit_dict=None, unit=None, filetype=None, pop_key=True
+        self,
+        ext_unit_dict: Union[str, os.PathLike] = None,
```
good catch, thanks, not sure how that slipped in

w-bonelli commented:
@mwtoews thanks for the catches, I think they've all been fixed.

mwtoews commented on flopy/mbase.py:
```diff
@@ -1499,13 +1517,15 @@ def add_pop_key_list(self, key):
 
-    def check(self, f=None, verbose=True, level=1):
+    def check(
+        self, f: Union[str, os.PathLike, None] = None, verbose=True, level=1
```
```suggestion
        self, f: Optional[Union[str, os.PathLike]] = None, verbose=True, level=1
```

mwtoews commented on flopy/mbase.py:
```diff
@@ -1499,13 +1517,15 @@ def add_pop_key_list(self, key):
         Parameters
         ----------
-        f : str or file handle
+        f : str or PathLike
```
```suggestion
        f : str or PathLike, optional
```

mwtoews commented on flopy/mbase.py:
```diff
@@ -1634,61 +1656,58 @@ def to_shapefile(self, filename, package_names=None, **kwargs):
-    model_ws="./",
-    scrub=False,
+    exe_name: Union[str, os.PathLike],
+    namefile: Optional[str],
```
Currently this is a required parameter, because it does not have a default e.g. `= None`.

w-bonelli replied:
The docstring indicated `namefile=None` could be used for models that don't require namefiles &mdash; the function seems to handle this case OK, just omits it from the args passed to the executable

mwtoews commented on flopy/export/shapefile_utils.py:
```diff
@@ -521,10 +531,10 @@ def shp2recarray(shpname):
     mg=None,
     epsg=None,
-    prj=None,
+    prj: Union[str, os.PathLike] = None,
```
```suggestion
    prj: Optional[Union[str, os.PathLike]] = None,
```

mwtoews commented on flopy/export/shapefile_utils.py:
```diff
@@ -620,7 +630,12 @@ def recarray2shp(
+    shpname: Union[str, os.PathLike],
+    mg=None,
+    epsg=None,
+    prj: Union[str, os.PathLike] = None,
```
```suggestion
    prj: Optional[Union[str, os.PathLike]] = None,
```

mwtoews reviewed (commented):
Just a few (probably) final comments. And again, I'm only reviewing the changes in `flopy/*`

w-bonelli commented:
Fixed more docstrings/typehints, thanks @mwtoews, also updated exe resolve function introduced in #1727
  - add docstrings
  - test cases [described](https://github.com/modflowpy/flopy/issues/1633#issuecomment-1396163755) in #1633
  - log externally (factor `silent` and `model_ws` params out)
  - name `resolve_exe_name` -> `resolve_exe` (full path is returned if successful)

mwtoews reviewed (approved):
Thanks, looks good! I only have a minor suggestion remaining. This PR could be merged soon.
